sunbeam-python/sunbeam/features/consul/feature.py

```diff
@@ -0,0 +1,554 @@
+        super().__init__()
+        self.tfplan_consul_client = CONSUL_CLIENT_TFPLAN
+        self.tfplan_consul_client_dir = "deploy-consul-client"
+
+    def default_software_overrides(self) -> SoftwareConfig:
+        """Manifest pluing part in dict format."""
+        return SoftwareConfig(
+            charms={
+                "consul-k8s": CharmManifest(channel=CONSUL_CHANNEL),
+                "consul-client": CharmManifest(channel=CONSUL_CHANNEL),
+            },
+            terraform={
+                self.tfplan_consul_client: TerraformManifest(
+                    source=Path(__file__).parent
+                    / "etc"  # noqa: W503
+                    / self.tfplan_consul_client_dir  # noqa: W503
+                ),
+            },
+        )
+
+    def manifest_attributes_tfvar_map(self) -> dict:
+        """Manifest attrbitues to terraformvars map."""
+        return {
+            self.tfplan: {
+                "charms": {
+                    "consul-k8s": {
+                        "channel": "consul-channel",
+                        "revision": "consul-revision",
+                        "config": "consul-config",
+                        "config-map": "consul-config-map",
+                    }
+                }
+            },
+            self.tfplan_consul_client: {
+                "charms": {
+                    "consul-client": {
+                        "channel": "consul-channel",
+                        "revision": "consul-revision",
+                        "config": "consul-config",
+                        "config-map": "consul-config-map",
+                    }
+                }
+            },
+        }
+
+    def consul_servers_to_enable(
+        self, deployment: Deployment
+    ) -> dict[ConsulServerNetworks, bool]:
+        """Return consul servers to enable.
+
+        Return dict to enable/disable consul server per network.
+        """
+        # Default to false
+        enable = dict.fromkeys(ConsulServerNetworks, False)
+
+        try:
+            management_space = deployment.get_space(Networks.MANAGEMENT)
+            enable[ConsulServerNetworks.MANAGEMENT] = True
+        except ValueError:
+            management_space = None
+
+        # If storage space is same as management space, dont enable consul
+        # server for storage
+        try:
+            storage_space = deployment.get_space(Networks.STORAGE)
+            if storage_space != management_space:
+                enable[ConsulServerNetworks.STORAGE] = True
+        except ValueError:
+            storage_space = None
+
+        # If data space is same as either of management or storage space,
+        # dont enable consul server for tenant
+        try:
+            tenant_space = deployment.get_space(Networks.DATA)
+            if tenant_space not in (management_space, storage_space):
+                enable[ConsulServerNetworks.TENANT] = True
+        except ValueError:
+            tenant_space = None
```

Comment:
This is just nitpicking, but I wonder if a client wouldn't find it strange if he looked closer and saw that he had a consul for management and one for storage, but none for tenant?

The same is true if he saw one for tenant but none for storage, but I would find that less shocking?

Reply:
I made this choice based on masakari default matrix.yaml for recovery mechanism where if storage is down, masakari triggers recovery but not with tenant. So i tried to prioritise consul cluster to run on storage when storage/tenant network space are same.

https://opendev.org/openstack/masakari-monitors/src/branch/master/etc/masakarimonitors/matrix.yaml.sample

Reply:
Ok, thank you! that's fine by me